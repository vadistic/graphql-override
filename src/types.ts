import {
  DirectiveDefinitionNode,
  DirectiveNode,
  EnumValueDefinitionNode,
  FieldDefinitionNode,
  InputValueDefinitionNode,
  Location,
  NamedTypeNode,
  NameNode,
  StringValueNode,
  TypeDefinitionNode,
  TypeExtensionNode,
} from 'graphql'
import { typeSystemHashMap } from './hashmap';

export type Mutable<T> = { -readonly [P in keyof T]: T[P] }

export type SupportedDefinitionNode =
  | TypeDefinitionNode
  | TypeExtensionNode
  | DirectiveDefinitionNode

// Types for keys of DefinitionNode props

type KeysOfUnion<T> = T extends any ? keyof T : never

export type AllDefinitionNodeKeys = KeysOfUnion<SupportedDefinitionNode>

export type RegularDefinitionNodeKeys =
  | 'kind'
  | 'loc'
  | 'description'
  | 'name'
  | 'locations'

export type HashDefinitionNodeKeys =
  | 'directives'
  | 'interfaces'
  | 'fields'
  | 'types'
  | 'values'
  | 'arguments'

export interface Hash<T> {
  [name: string]: T
}

export interface HashedDefinitionNode<
  T extends SupportedDefinitionNode = SupportedDefinitionNode
> {
  kind: keyof typeof typeSystemHashMap
  loc?: Location
  description?: 'description' extends keyof T ? StringValueNode : never
  name?: NameNode
  interfaces?: 'interfaces' extends keyof T ? Hash<NamedTypeNode> : never
  types?: 'types' extends keyof T ? Hash<NamedTypeNode> : never
  directives?: 'directives' extends keyof T ? Hash<DirectiveNode> : never
  fields?: 'fields' extends keyof T
    ? Hash<FieldDefinitionNode> | Hash<InputValueDefinitionNode>
    : never
  values?: 'values' extends keyof T ? Hash<EnumValueDefinitionNode> : never
  arguments?: 'arguments' extends keyof T ? Hash<InputValueDefinitionNode> : never
  // not hashing because the name value is missing
  locations: 'locations' extends keyof T ? ReadonlyArray<NameNode> : never
}
