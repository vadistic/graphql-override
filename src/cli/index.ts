#!/usr/bin/env node
import * as meow from 'meow'

import { cli } from './cli'
export * from './cli'

export const main = meow(
  `
  Usage
  $ graphql-override <flags>

  Options
  --config, -c      path to .graphqlconfig (if not in root dir)
  --schema, -s      path to main schema
  --overrides, -r   path to overrides
  --output, -o      location for generated file (with filename and extension)

  Example with .graphqlconfig
  $ graphql-override

  Example without .graphql config (all flags required)
  $ graphql-override -s src/schema/schema.graphql -o src/schema/overrides.graphql -o src/generated/app.graphql

  <cool emoji>
  `,
  {
    flags: {
      config: {
        type: 'string',
        alias: 'c',
      },
      schema: {
        type: 'string',
        alias: 's',
      },
      overrides: {
        type: 'string',
        alias: 'r',
      },
      output: {
        type: 'string',
        alias: 'o',
      },
    },
  }
)

// run!
cli(main.flags)
