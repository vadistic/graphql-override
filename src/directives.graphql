# Directives for TYPE_DEFINITION | FIELD DEFINITION
directive @create on FIELD_DEFINITION
directive @replace on FIELD_DEFINITION
directive @upsert on FIELD_DEFINITION
directive @delete on FIELD_DEFINITION
directive @remove on FIELD_DEFINITION

# Directives only for TYPE_DEFINITION
directive @extend on FIELD_DEFINITION
directive @exclude on FIELD_DEFINITION
