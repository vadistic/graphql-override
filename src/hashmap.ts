// TypeDefinitionNode
export const typeDefinitionHashMap = {
  ScalarTypeDefinition: {
    kind: false,
    loc: false,
    description: false,
    name: false,
    directives: true,
  },
  ObjectTypeDefinition: {
    kind: false,
    loc: false,
    description: false,
    name: false,
    interfaces: true,
    directives: true,
    fields: true,
  },
  InterfaceTypeDefinition: {
    kind: false,
    loc: false,
    description: false,
    name: false,
    directives: true,
    fields: true,
  },
  UnionTypeDefinition: {
    kind: false,
    loc: false,
    description: false,
    name: false,
    directives: true,
  },
  EnumTypeDefinition: {
    kind: false,
    loc: false,
    description: false,
    name: false,
    directives: true,
    values: true,
  },
  InputObjectTypeDefinition: {
    kind: false,
    loc: false,
    description: false,
    name: false,
    directives: true,
    fields: true,
  },
}

// TypeExtensionNode
export const typeExtensionHashMap = {
  ScalarTypeExtension: {
    kind: false,
    loc: false,
    name: false,
    directives: true,
  },
  ObjectTypeExtension: {
    kind: false,
    loc: false,
    name: false,
    interfaces: true,
    directives: true,
    fields: true,
  },
  InterfaceTypeExtension: {
    kind: false,
    loc: false,
    name: false,
    directives: true,
    fields: true,
  },
  UnionTypeExtension: {
    kind: false,
    loc: false,
    name: false,
    directives: true,
  },
  EnumTypeExtension: {
    kind: false,
    loc: false,
    name: false,
    directives: true,
    values: true,
  },
  InputObjectTypeExtension: {
    kind: false,
    loc: false,
    name: false,
    directives: true,
    fields: true,
  },
}

// DirectiveDefinitionNode
export const directiveDefinitionHashMap = {
  DirectiveDefinition: {
    kind: false,
    loc: false,
    description: false,
    name: false,
    arguments: true,
    // Is Array, but has not name.value so cannot hash it with current system
    locations: false,
  },
}

export const typeSystemHashMap = {
  ...typeDefinitionHashMap,
  ...typeExtensionHashMap,
  ...directiveDefinitionHashMap,
}
