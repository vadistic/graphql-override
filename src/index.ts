export * from './DefinitionEditor'
export * from './TypeDefsEditor'
export * from './override'
