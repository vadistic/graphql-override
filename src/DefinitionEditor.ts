import {
  ASTKindToNode,
  DefinitionNode,
  DirectiveNode,
  EnumValueDefinitionNode,
  FieldDefinitionNode,
  InputValueDefinitionNode,
  NamedTypeNode,
  print,
} from 'graphql'
import * as R from 'ramda'

import { definitionHashMap } from './hashmap'
import {
  AllDefinitionNodeKeys,
  HashDefinitionNodeKeys,
  HashedDefinitionNode,
  SupportedDefinitionNode,
} from './types'
import { hashArrByName, isSupported } from './util'

export type HasProp = (prop: HashDefinitionNodeKeys) => (name: string) => boolean

export type ExistInProp = (prop: HashDefinitionNodeKeys) => (name: string) => boolean

export interface GetInProp {
  (prop: 'arguments'): (name: string) => InputValueDefinitionNode
  (prop: 'directives'): (name: string) => DirectiveNode
  (prop: 'fields'): (name: string) => FieldDefinitionNode | InputValueDefinitionNode
  (prop: 'interfaces' | 'types'): (name: string) => NamedTypeNode
  (prop: 'values'): (name: string) => EnumValueDefinitionNode
}

export interface CreateOnType<T extends SupportedDefinitionNode> {
  (prop: 'arguments'): (def: InputValueDefinitionNode) => GraphqlDefinitionEditor<T>
  (prop: 'directives'): (def: DirectiveNode) => GraphqlDefinitionEditor<T>
  (prop: 'fields'): (
    def: FieldDefinitionNode | InputValueDefinitionNode
  ) => GraphqlDefinitionEditor<T>
  (prop: 'interfaces'): (def: NamedTypeNode) => GraphqlDefinitionEditor<T>
  (prop: 'values'): (def: EnumValueDefinitionNode) => GraphqlDefinitionEditor<T>
}

export type UpsertOnType<T extends SupportedDefinitionNode> = CreateOnType<T>

export type DeleteOnType<T extends SupportedDefinitionNode> = (
  prop: HashDefinitionNodeKeys
) => (name: string) => GraphqlDefinitionEditor<T>

export class GraphqlDefinitionEditor<
  T extends SupportedDefinitionNode = SupportedDefinitionNode
> {
  // Type assertion for object initialization
  private hashNode = {} as HashedDefinitionNode<T>

  constructor(node: SupportedDefinitionNode) {
    if (!isSupported(node)) {
      // Type assertion because error is for non-TS / invalid typings
      throw Error(`Type ${node!.kind} is not supported by  type definition editor`)
    }

    // Hashing all array properties using name as key
    Object.keys(node).forEach(prop => {
      if (definitionHashMap[node.kind][prop]) {
        this.hashNode[prop] = hashArrByName(node[prop])
      } else {
        this.hashNode[prop] = node[prop]
      }
    })
  }

  // Helper
  public name = () => this.hashNode.name.value
  public kind = () => this.hashNode.kind

  public hasProp: HasProp = prop => name => R.has(name, this.hashNode)

  public existInProp: ExistInProp = prop => name => R.has(name, this.hashNode[prop])

  public getInProp: GetInProp = prop => name => R.prop(name, this.hashNode[prop])

  public createInProp: CreateOnType<T> = prop => def => {
    if (!R.has(prop, this.hashNode)) {
      throw new Error(
        `Cannot create ${def.name.value} in ${prop} on type ${this.name()}
        ${this.kind()} type has no ${prop}.
        `
      )
    }

    if (this.existInProp(prop)(def.name.value)) {
      throw new Error(
        `Cannot create ${def.name.value} in ${prop} on type ${this.name()}
        Value with the same name already exist.
        `
      )
    }

    this.hashNode[prop][def.name.value] = def

    return this
  }

  public upsertInProp: UpsertOnType<T> = prop => def => {
    if (!R.has(prop, this.hashNode)) {
      throw new Error(
        `Cannot create ${def.name.value} in ${prop} on type ${this.name()}
        ${this.kind()} type has no ${prop}.
        `
      )
    }

    this.hashNode[prop][def.name.value] = def

    return this
  }

  public deleteInProp: DeleteOnType<T> = prop => name => {
    if (!R.has(prop, this.hashNode)) {
      throw new Error(
        `Cannot delete ${name} in ${prop} on type ${this.name()}
        ${this.kind()} type has no ${prop}.
        `
      )
    }

    if (!this.existInProp(prop)(name)) {
      throw new Error(
        `Cannot delete ${name} in ${prop} on type ${this.name()}
        Value does not exist.
        `
      )
    }

    this.hashNode[prop] = R.dissoc(name, this.hashNode[prop])

    return this
  }

  // Crud
  public existDirective = this.existInProp('directives')
  public existField = this.existInProp('fields')
  public existInterface = this.existInProp('interfaces')
  public existValue = this.existInProp('values')

  public getDirective = this.getInProp('directives')
  public getField = this.getInProp('fields')
  public getInterface = this.getInProp('interfaces')
  public getValue = this.getInProp('values')

  public createDirective = this.createInProp('directives')
  public createField = this.createInProp('fields')
  public createInterface = this.createInProp('interfaces')
  public createValue = this.createInProp('values')

  public upsertDirective = this.upsertInProp('directives')
  public upsertField = this.upsertInProp('fields')
  public upsertInterface = this.upsertInProp('interfaces')
  public upsertValue = this.upsertInProp('values')

  public deleteDirective = this.deleteInProp('directives')
  public deleteField = this.deleteInProp('fields')
  public deleteInterface = this.deleteInProp('interfaces')
  public deleteValue = this.deleteInProp('values')

  public astNode = () => {
    const astNode = Object.keys(this.hashNode).reduce(
      (acc, prop: AllDefinitionNodeKeys) => {
        if (definitionHashMap[this.kind()][prop]) {
          acc[prop] = Object.values(this.hashNode[prop])
        } else {
          acc[prop] = this.hashNode[prop]
        }
        return acc
      },
      {}
    )
    return astNode as SupportedDefinitionNode
  }

  public print = () => print(this.astNode())
}
