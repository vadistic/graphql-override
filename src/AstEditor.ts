import {
  DefinitionNode,
  DocumentNode,
  EnumValueDefinitionNode,
  FieldDefinitionNode,
  InterfaceTypeDefinitionNode,
  Location,
  parse,
  print,
  TypeDefinitionNode,
} from 'graphql'
import * as R from 'ramda'

import { GraphqlDefinitionEditor } from './DefinitionEditor'
import { definitionHashMap } from './hashmap'
import { Hash, SupportedDefinitionNode } from './types'
import { hashArrByName, hasName, isSupported } from './util'

export class GraphqlAstEditor {
  public hashDefs: Hash<GraphqlDefinitionEditor> = {}
  private schema: DocumentNode

  constructor(schema: string | DocumentNode) {
    this.schema = typeof schema === 'string' ? parse(schema) : schema

    this.schema.definitions.forEach(node => {
      if (!isSupported(node)) {
        throw Error(`Type ${node.kind} is not supported by type definition editor`)
      } else {
        this.hashDefs[node.name.value] = new GraphqlDefinitionEditor(node)
      }
    })
  }

  public astNode = () => ({
    ...this.schema,
    definitions: Object.values(this.hashDefs).reduce((acc, def) => {
      acc.push(def.astNode())
      return acc
    }, []),
  })

  public print = () => print(this.astNode())
}
