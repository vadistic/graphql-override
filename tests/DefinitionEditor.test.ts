import { parse } from 'graphql'
import * as prettier from 'prettier'

import { GraphqlDefinitionEditor } from '../src'
import { SupportedDefinitionNode } from '../src/types'

const pretty = (source: string) => prettier.format(source, { parser: 'graphql' })

const serialize = (obj: object) => JSON.stringify(obj, null, 2)

const getDef = (source: string) => parse(source).definitions[0] as SupportedDefinitionNode

describe('DefinitionEditor SupportedNodeTypes', () => {
  it('parse & print ScalarTypeDefinition', () => {
    const source = pretty(`
    scalar Date
    `)

    const editor = new GraphqlDefinitionEditor(getDef(source))

    expect(editor.kind()).toMatch('ScalarTypeDefinition')
    expect(pretty(editor.print())).toMatch(source)
    expect(serialize(editor.node())).toMatchSnapshot()
  })

  it('parse, print & returns node of ObjectTypeDefinition', () => {
    const source = pretty(`
    type Query {
      Tweet(id: ID!): Tweet
      Tweets(limit: Int, skip: Int, sort_field: String, sort_order: String): [Tweet]
    }
    `)

    const editor = new GraphqlDefinitionEditor(getDef(source))

    expect(editor.kind()).toMatch('ObjectTypeDefinition')
    expect(pretty(editor.print())).toMatch(source)
    expect(serialize(editor.node())).toMatchSnapshot()
  })

  it('parse, print & returns node of InterfaceTypeDefinition', () => {
    const source = pretty(`
    interface BssicNode {
      id: ID!
      createdAt: DateTime!
      deletedAt: DateTime
    }
    `)

    const editor = new GraphqlDefinitionEditor(getDef(source))

    expect(editor.kind()).toMatch('InterfaceTypeDefinition')
    expect(pretty(editor.print())).toMatch(source)
    expect(serialize(editor.node())).toMatchSnapshot()
  })

  it('parse, print & returns node of UnionTypeDefinition', () => {
    const source = pretty(`
    union Person = User | Candidate
    `)

    const editor = new GraphqlDefinitionEditor(getDef(source))

    expect(editor.kind()).toMatch('UnionTypeDefinition')
    expect(pretty(editor.print())).toMatch(source)
    expect(serialize(editor.node())).toMatchSnapshot()
  })

  it('parse & print & returns node of EnumTypeDefinition', () => {
    const source = pretty(`
    enum OrderBy {
      name
      age
      last_login
    }
    `)

    const editor = new GraphqlDefinitionEditor(getDef(source))

    expect(editor.kind()).toMatch('EnumTypeDefinition')
    expect(pretty(editor.print())).toMatch(source)
    expect(serialize(editor.node())).toMatchSnapshot()
  })

  it('parse, print & returns node of InputObjectTypeDefinition', () => {
    const source = pretty(`
    input CreateTweetInput {
      body: String
      tags: [String]!
    }
    `)

    const editor = new GraphqlDefinitionEditor(getDef(source))

    expect(editor.kind()).toMatch('InputObjectTypeDefinition')
    expect(pretty(editor.print())).toMatch(source)
    expect(serialize(editor.node())).toMatchSnapshot()
  })

  it('parse, print & returns node of DirectiveDefinition', () => {
    const source = pretty(`
    directive @deprecated(
      reason: String = "No longer supported"
    ) on FIELD_DEFINITION | ENUM_VALUE
    `)

    const editor = new GraphqlDefinitionEditor(getDef(source))

    expect(editor.kind()).toMatch('DirectiveDefinition')
    expect(pretty(editor.print())).toMatch(source)
    expect(serialize(editor.node())).toMatchSnapshot()
  })
})
