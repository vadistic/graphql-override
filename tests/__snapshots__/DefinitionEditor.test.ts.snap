// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`DefinitionEditor SupportedNodeTypes parse & print & returns node of EnumTypeDefinition 1`] = `
"{
  \\"kind\\": \\"EnumTypeDefinition\\",
  \\"name\\": {
    \\"kind\\": \\"Name\\",
    \\"value\\": \\"OrderBy\\",
    \\"loc\\": {
      \\"start\\": 5,
      \\"end\\": 12
    }
  },
  \\"directives\\": [],
  \\"values\\": [
    {
      \\"kind\\": \\"EnumValueDefinition\\",
      \\"name\\": {
        \\"kind\\": \\"Name\\",
        \\"value\\": \\"name\\",
        \\"loc\\": {
          \\"start\\": 17,
          \\"end\\": 21
        }
      },
      \\"directives\\": [],
      \\"loc\\": {
        \\"start\\": 17,
        \\"end\\": 21
      }
    },
    {
      \\"kind\\": \\"EnumValueDefinition\\",
      \\"name\\": {
        \\"kind\\": \\"Name\\",
        \\"value\\": \\"age\\",
        \\"loc\\": {
          \\"start\\": 24,
          \\"end\\": 27
        }
      },
      \\"directives\\": [],
      \\"loc\\": {
        \\"start\\": 24,
        \\"end\\": 27
      }
    },
    {
      \\"kind\\": \\"EnumValueDefinition\\",
      \\"name\\": {
        \\"kind\\": \\"Name\\",
        \\"value\\": \\"last_login\\",
        \\"loc\\": {
          \\"start\\": 30,
          \\"end\\": 40
        }
      },
      \\"directives\\": [],
      \\"loc\\": {
        \\"start\\": 30,
        \\"end\\": 40
      }
    }
  ],
  \\"loc\\": {
    \\"start\\": 0,
    \\"end\\": 42
  }
}"
`;

exports[`DefinitionEditor SupportedNodeTypes parse & print ScalarTypeDefinition 1`] = `
"{
  \\"kind\\": \\"ScalarTypeDefinition\\",
  \\"name\\": {
    \\"kind\\": \\"Name\\",
    \\"value\\": \\"Date\\",
    \\"loc\\": {
      \\"start\\": 7,
      \\"end\\": 11
    }
  },
  \\"directives\\": [],
  \\"loc\\": {
    \\"start\\": 0,
    \\"end\\": 11
  }
}"
`;

exports[`DefinitionEditor SupportedNodeTypes parse, print & returns node of DirectiveDefinition 1`] = `
"{
  \\"kind\\": \\"DirectiveDefinition\\",
  \\"name\\": {
    \\"kind\\": \\"Name\\",
    \\"value\\": \\"deprecated\\",
    \\"loc\\": {
      \\"start\\": 11,
      \\"end\\": 21
    }
  },
  \\"arguments\\": [
    {
      \\"kind\\": \\"InputValueDefinition\\",
      \\"name\\": {
        \\"kind\\": \\"Name\\",
        \\"value\\": \\"reason\\",
        \\"loc\\": {
          \\"start\\": 25,
          \\"end\\": 31
        }
      },
      \\"type\\": {
        \\"kind\\": \\"NamedType\\",
        \\"name\\": {
          \\"kind\\": \\"Name\\",
          \\"value\\": \\"String\\",
          \\"loc\\": {
            \\"start\\": 33,
            \\"end\\": 39
          }
        },
        \\"loc\\": {
          \\"start\\": 33,
          \\"end\\": 39
        }
      },
      \\"defaultValue\\": {
        \\"kind\\": \\"StringValue\\",
        \\"value\\": \\"No longer supported\\",
        \\"block\\": false,
        \\"loc\\": {
          \\"start\\": 42,
          \\"end\\": 63
        }
      },
      \\"directives\\": [],
      \\"loc\\": {
        \\"start\\": 25,
        \\"end\\": 63
      }
    }
  ],
  \\"locations\\": [
    {
      \\"kind\\": \\"Name\\",
      \\"value\\": \\"FIELD_DEFINITION\\",
      \\"loc\\": {
        \\"start\\": 69,
        \\"end\\": 85
      }
    },
    {
      \\"kind\\": \\"Name\\",
      \\"value\\": \\"ENUM_VALUE\\",
      \\"loc\\": {
        \\"start\\": 88,
        \\"end\\": 98
      }
    }
  ],
  \\"loc\\": {
    \\"start\\": 0,
    \\"end\\": 98
  }
}"
`;

exports[`DefinitionEditor SupportedNodeTypes parse, print & returns node of InputObjectTypeDefinition 1`] = `
"{
  \\"kind\\": \\"InputObjectTypeDefinition\\",
  \\"name\\": {
    \\"kind\\": \\"Name\\",
    \\"value\\": \\"CreateTweetInput\\",
    \\"loc\\": {
      \\"start\\": 6,
      \\"end\\": 22
    }
  },
  \\"directives\\": [],
  \\"fields\\": [
    {
      \\"kind\\": \\"InputValueDefinition\\",
      \\"name\\": {
        \\"kind\\": \\"Name\\",
        \\"value\\": \\"body\\",
        \\"loc\\": {
          \\"start\\": 27,
          \\"end\\": 31
        }
      },
      \\"type\\": {
        \\"kind\\": \\"NamedType\\",
        \\"name\\": {
          \\"kind\\": \\"Name\\",
          \\"value\\": \\"String\\",
          \\"loc\\": {
            \\"start\\": 33,
            \\"end\\": 39
          }
        },
        \\"loc\\": {
          \\"start\\": 33,
          \\"end\\": 39
        }
      },
      \\"directives\\": [],
      \\"loc\\": {
        \\"start\\": 27,
        \\"end\\": 39
      }
    },
    {
      \\"kind\\": \\"InputValueDefinition\\",
      \\"name\\": {
        \\"kind\\": \\"Name\\",
        \\"value\\": \\"tags\\",
        \\"loc\\": {
          \\"start\\": 42,
          \\"end\\": 46
        }
      },
      \\"type\\": {
        \\"kind\\": \\"NonNullType\\",
        \\"type\\": {
          \\"kind\\": \\"ListType\\",
          \\"type\\": {
            \\"kind\\": \\"NamedType\\",
            \\"name\\": {
              \\"kind\\": \\"Name\\",
              \\"value\\": \\"String\\",
              \\"loc\\": {
                \\"start\\": 49,
                \\"end\\": 55
              }
            },
            \\"loc\\": {
              \\"start\\": 49,
              \\"end\\": 55
            }
          },
          \\"loc\\": {
            \\"start\\": 48,
            \\"end\\": 56
          }
        },
        \\"loc\\": {
          \\"start\\": 48,
          \\"end\\": 57
        }
      },
      \\"directives\\": [],
      \\"loc\\": {
        \\"start\\": 42,
        \\"end\\": 57
      }
    }
  ],
  \\"loc\\": {
    \\"start\\": 0,
    \\"end\\": 59
  }
}"
`;

exports[`DefinitionEditor SupportedNodeTypes parse, print & returns node of InterfaceTypeDefinition 1`] = `
"{
  \\"kind\\": \\"InterfaceTypeDefinition\\",
  \\"name\\": {
    \\"kind\\": \\"Name\\",
    \\"value\\": \\"BssicNode\\",
    \\"loc\\": {
      \\"start\\": 10,
      \\"end\\": 19
    }
  },
  \\"directives\\": [],
  \\"fields\\": [
    {
      \\"kind\\": \\"FieldDefinition\\",
      \\"name\\": {
        \\"kind\\": \\"Name\\",
        \\"value\\": \\"id\\",
        \\"loc\\": {
          \\"start\\": 24,
          \\"end\\": 26
        }
      },
      \\"arguments\\": [],
      \\"type\\": {
        \\"kind\\": \\"NonNullType\\",
        \\"type\\": {
          \\"kind\\": \\"NamedType\\",
          \\"name\\": {
            \\"kind\\": \\"Name\\",
            \\"value\\": \\"ID\\",
            \\"loc\\": {
              \\"start\\": 28,
              \\"end\\": 30
            }
          },
          \\"loc\\": {
            \\"start\\": 28,
            \\"end\\": 30
          }
        },
        \\"loc\\": {
          \\"start\\": 28,
          \\"end\\": 31
        }
      },
      \\"directives\\": [],
      \\"loc\\": {
        \\"start\\": 24,
        \\"end\\": 31
      }
    },
    {
      \\"kind\\": \\"FieldDefinition\\",
      \\"name\\": {
        \\"kind\\": \\"Name\\",
        \\"value\\": \\"createdAt\\",
        \\"loc\\": {
          \\"start\\": 34,
          \\"end\\": 43
        }
      },
      \\"arguments\\": [],
      \\"type\\": {
        \\"kind\\": \\"NonNullType\\",
        \\"type\\": {
          \\"kind\\": \\"NamedType\\",
          \\"name\\": {
            \\"kind\\": \\"Name\\",
            \\"value\\": \\"DateTime\\",
            \\"loc\\": {
              \\"start\\": 45,
              \\"end\\": 53
            }
          },
          \\"loc\\": {
            \\"start\\": 45,
            \\"end\\": 53
          }
        },
        \\"loc\\": {
          \\"start\\": 45,
          \\"end\\": 54
        }
      },
      \\"directives\\": [],
      \\"loc\\": {
        \\"start\\": 34,
        \\"end\\": 54
      }
    },
    {
      \\"kind\\": \\"FieldDefinition\\",
      \\"name\\": {
        \\"kind\\": \\"Name\\",
        \\"value\\": \\"deletedAt\\",
        \\"loc\\": {
          \\"start\\": 57,
          \\"end\\": 66
        }
      },
      \\"arguments\\": [],
      \\"type\\": {
        \\"kind\\": \\"NamedType\\",
        \\"name\\": {
          \\"kind\\": \\"Name\\",
          \\"value\\": \\"DateTime\\",
          \\"loc\\": {
            \\"start\\": 68,
            \\"end\\": 76
          }
        },
        \\"loc\\": {
          \\"start\\": 68,
          \\"end\\": 76
        }
      },
      \\"directives\\": [],
      \\"loc\\": {
        \\"start\\": 57,
        \\"end\\": 76
      }
    }
  ],
  \\"loc\\": {
    \\"start\\": 0,
    \\"end\\": 78
  }
}"
`;

exports[`DefinitionEditor SupportedNodeTypes parse, print & returns node of ObjectTypeDefinition 1`] = `
"{
  \\"kind\\": \\"ObjectTypeDefinition\\",
  \\"name\\": {
    \\"kind\\": \\"Name\\",
    \\"value\\": \\"Query\\",
    \\"loc\\": {
      \\"start\\": 5,
      \\"end\\": 10
    }
  },
  \\"interfaces\\": [],
  \\"directives\\": [],
  \\"fields\\": [
    {
      \\"kind\\": \\"FieldDefinition\\",
      \\"name\\": {
        \\"kind\\": \\"Name\\",
        \\"value\\": \\"Tweet\\",
        \\"loc\\": {
          \\"start\\": 15,
          \\"end\\": 20
        }
      },
      \\"arguments\\": [
        {
          \\"kind\\": \\"InputValueDefinition\\",
          \\"name\\": {
            \\"kind\\": \\"Name\\",
            \\"value\\": \\"id\\",
            \\"loc\\": {
              \\"start\\": 21,
              \\"end\\": 23
            }
          },
          \\"type\\": {
            \\"kind\\": \\"NonNullType\\",
            \\"type\\": {
              \\"kind\\": \\"NamedType\\",
              \\"name\\": {
                \\"kind\\": \\"Name\\",
                \\"value\\": \\"ID\\",
                \\"loc\\": {
                  \\"start\\": 25,
                  \\"end\\": 27
                }
              },
              \\"loc\\": {
                \\"start\\": 25,
                \\"end\\": 27
              }
            },
            \\"loc\\": {
              \\"start\\": 25,
              \\"end\\": 28
            }
          },
          \\"directives\\": [],
          \\"loc\\": {
            \\"start\\": 21,
            \\"end\\": 28
          }
        }
      ],
      \\"type\\": {
        \\"kind\\": \\"NamedType\\",
        \\"name\\": {
          \\"kind\\": \\"Name\\",
          \\"value\\": \\"Tweet\\",
          \\"loc\\": {
            \\"start\\": 31,
            \\"end\\": 36
          }
        },
        \\"loc\\": {
          \\"start\\": 31,
          \\"end\\": 36
        }
      },
      \\"directives\\": [],
      \\"loc\\": {
        \\"start\\": 15,
        \\"end\\": 36
      }
    },
    {
      \\"kind\\": \\"FieldDefinition\\",
      \\"name\\": {
        \\"kind\\": \\"Name\\",
        \\"value\\": \\"Tweets\\",
        \\"loc\\": {
          \\"start\\": 39,
          \\"end\\": 45
        }
      },
      \\"arguments\\": [
        {
          \\"kind\\": \\"InputValueDefinition\\",
          \\"name\\": {
            \\"kind\\": \\"Name\\",
            \\"value\\": \\"limit\\",
            \\"loc\\": {
              \\"start\\": 46,
              \\"end\\": 51
            }
          },
          \\"type\\": {
            \\"kind\\": \\"NamedType\\",
            \\"name\\": {
              \\"kind\\": \\"Name\\",
              \\"value\\": \\"Int\\",
              \\"loc\\": {
                \\"start\\": 53,
                \\"end\\": 56
              }
            },
            \\"loc\\": {
              \\"start\\": 53,
              \\"end\\": 56
            }
          },
          \\"directives\\": [],
          \\"loc\\": {
            \\"start\\": 46,
            \\"end\\": 56
          }
        },
        {
          \\"kind\\": \\"InputValueDefinition\\",
          \\"name\\": {
            \\"kind\\": \\"Name\\",
            \\"value\\": \\"skip\\",
            \\"loc\\": {
              \\"start\\": 58,
              \\"end\\": 62
            }
          },
          \\"type\\": {
            \\"kind\\": \\"NamedType\\",
            \\"name\\": {
              \\"kind\\": \\"Name\\",
              \\"value\\": \\"Int\\",
              \\"loc\\": {
                \\"start\\": 64,
                \\"end\\": 67
              }
            },
            \\"loc\\": {
              \\"start\\": 64,
              \\"end\\": 67
            }
          },
          \\"directives\\": [],
          \\"loc\\": {
            \\"start\\": 58,
            \\"end\\": 67
          }
        },
        {
          \\"kind\\": \\"InputValueDefinition\\",
          \\"name\\": {
            \\"kind\\": \\"Name\\",
            \\"value\\": \\"sort_field\\",
            \\"loc\\": {
              \\"start\\": 69,
              \\"end\\": 79
            }
          },
          \\"type\\": {
            \\"kind\\": \\"NamedType\\",
            \\"name\\": {
              \\"kind\\": \\"Name\\",
              \\"value\\": \\"String\\",
              \\"loc\\": {
                \\"start\\": 81,
                \\"end\\": 87
              }
            },
            \\"loc\\": {
              \\"start\\": 81,
              \\"end\\": 87
            }
          },
          \\"directives\\": [],
          \\"loc\\": {
            \\"start\\": 69,
            \\"end\\": 87
          }
        },
        {
          \\"kind\\": \\"InputValueDefinition\\",
          \\"name\\": {
            \\"kind\\": \\"Name\\",
            \\"value\\": \\"sort_order\\",
            \\"loc\\": {
              \\"start\\": 89,
              \\"end\\": 99
            }
          },
          \\"type\\": {
            \\"kind\\": \\"NamedType\\",
            \\"name\\": {
              \\"kind\\": \\"Name\\",
              \\"value\\": \\"String\\",
              \\"loc\\": {
                \\"start\\": 101,
                \\"end\\": 107
              }
            },
            \\"loc\\": {
              \\"start\\": 101,
              \\"end\\": 107
            }
          },
          \\"directives\\": [],
          \\"loc\\": {
            \\"start\\": 89,
            \\"end\\": 107
          }
        }
      ],
      \\"type\\": {
        \\"kind\\": \\"ListType\\",
        \\"type\\": {
          \\"kind\\": \\"NamedType\\",
          \\"name\\": {
            \\"kind\\": \\"Name\\",
            \\"value\\": \\"Tweet\\",
            \\"loc\\": {
              \\"start\\": 111,
              \\"end\\": 116
            }
          },
          \\"loc\\": {
            \\"start\\": 111,
            \\"end\\": 116
          }
        },
        \\"loc\\": {
          \\"start\\": 110,
          \\"end\\": 117
        }
      },
      \\"directives\\": [],
      \\"loc\\": {
        \\"start\\": 39,
        \\"end\\": 117
      }
    }
  ],
  \\"loc\\": {
    \\"start\\": 0,
    \\"end\\": 119
  }
}"
`;

exports[`DefinitionEditor SupportedNodeTypes parse, print & returns node of UnionTypeDefinition 1`] = `
"{
  \\"kind\\": \\"UnionTypeDefinition\\",
  \\"name\\": {
    \\"kind\\": \\"Name\\",
    \\"value\\": \\"Person\\",
    \\"loc\\": {
      \\"start\\": 6,
      \\"end\\": 12
    }
  },
  \\"directives\\": [],
  \\"types\\": [
    {
      \\"kind\\": \\"NamedType\\",
      \\"name\\": {
        \\"kind\\": \\"Name\\",
        \\"value\\": \\"User\\",
        \\"loc\\": {
          \\"start\\": 15,
          \\"end\\": 19
        }
      },
      \\"loc\\": {
        \\"start\\": 15,
        \\"end\\": 19
      }
    },
    {
      \\"kind\\": \\"NamedType\\",
      \\"name\\": {
        \\"kind\\": \\"Name\\",
        \\"value\\": \\"Candidate\\",
        \\"loc\\": {
          \\"start\\": 22,
          \\"end\\": 31
        }
      },
      \\"loc\\": {
        \\"start\\": 22,
        \\"end\\": 31
      }
    }
  ],
  \\"loc\\": {
    \\"start\\": 0,
    \\"end\\": 31
  }
}"
`;
